### PR TITLE
Remove Database validation in jdl

### LINKED ISSUE #12959: Customize database validation in JDL
##### **Overview of the issue**

Since database type is checked in import-jdl the dotnet blueprint can't import jdl for sqllite database (in-memory) 
-> https://github.com/jhipster/generator-jhipster/blob/main/jdl/jhipster/field-types.js#L144

##### **Suggest a Fix**

Maybe we can give the possibility to customize the supported database types and blueprints can enrich them

##### **Browsers and Operating System**

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by nicolas63**
---

Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [ ] Tests are added where necessary
-   [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

fix #12959

### CODE REVIEW

**Comment by MathieuAA** [comment]
Validation as a whole should be disabled when using blueprints.... Not just that.

**Comment by nicolas63** [comment]
sorry for my misunderstanding, how can I know if a blueprint is used in this part?

**Comment by MathieuAA** [comment]
@nicolas63 don't be! I'm gonna walk you through how it can be done step by step.

If you look at the `jdl/jdl-importer.js` file, you'll see it's the place where the actual JDL importing is done.
From there, there are two paths:
  - the first one for importing JDL without applications,
  - the second one reserved to the opposite: importing JDL with applications (at least one)

Now, go line 179. you'll see this function:

```javascript
function checkForErrors(jdlObject, configuration) {
    let validator;
    if (jdlObject.getApplicationQuantity() === 0) {
        let application = configuration.application;
        if (!application && doesFileExist('.yo-rc.json')) {
            application = readJSONFile('.yo-rc.json');
        }
        let applicationType = configuration.applicationType;
        let databaseType = configuration.databaseType;
        let skippedUserManagement = configuration.skipUserManagement;
        if (application && application['generator-jhipster']) {
            if (applicationType === undefined) {
                applicationType = application['generator-jhipster'].applicationType;
            }
            if (databaseType === undefined) {
                databaseType = application['generator-jhipster'].databaseType;
            }
            if (skippedUserManagement === undefined) {
                skippedUserManagement = application['generator-jhipster'].skipUserManagement;
            }
        }
        validator = JDLWithoutApplicationValidator.createValidator(jdlObject, {
            applicationType,
            databaseType,
            skippedUserManagement,
        });
    } else {
        validator = JDLWithApplicationValidator.createValidator(jdlObject);
    }
    validator.checkForErrors();
}
```
This function checks a passed JDL object for errors. This is the validation is done. When there's no blueprint, the actual rules for validations are implemented and it works.
When there are blueprints, one can augment the JHipster scope to add databases, field types, etc. but they will trigger an error from this function, as you noticed.

There are two fixes to do:
  - a first one for JDLs without apps,
  - another one for JDL with apps.

There are two files: `jdl/validators/jdl-without-application-validator.js` and `jdl/validators/jdl-with-validator.js` that we need fix.

For the "no app" case, we need to pass to the validator whether there are blueprints involved, which is done line 173 of the `jdl-importer.js` file where we need to add one more parameter to the applicationSettings: `blueprints: ...` like this:

```javascript
function checkForErrors(jdlObject, configuration) {
    let validator;
    if (jdlObject.getApplicationQuantity() === 0) {
        let application = configuration.application;
        if (!application && doesFileExist('.yo-rc.json')) {
            application = readJSONFile('.yo-rc.json');
        }
        let applicationType = configuration.applicationType;
        let databaseType = configuration.databaseType;
        let skippedUserManagement = configuration.skipUserManagement;
        let blueprints = configuration.blueprints; // first there
        if (application && application['generator-jhipster']) {
            if (applicationType === undefined) {
                applicationType = application['generator-jhipster'].applicationType;
            }
            if (databaseType === undefined) {
                databaseType = application['generator-jhipster'].databaseType;
            }
            if (skippedUserManagement === undefined) {
                skippedUserManagement = application['generator-jhipster'].skipUserManagement;
            }
            if (blueprints === undefined) { // here
                blueprints = application['generator-jhipster'].blueprints;
            }
        }
        validator = JDLWithoutApplicationValidator.createValidator(jdlObject, {
            applicationType,
            databaseType,
            skippedUserManagement,
            blueprints, // and there
        });
    } else {
        validator = JDLWithApplicationValidator.createValidator(jdlObject);
    }
    validator.checkForErrors();
}
```
and finally in the `jdl/validators/jdl-without-application-validator.js` file we need to:
  - add the new blueprints param to the documentation,
  - and add an if to the `createValidator` function like this:
```javascript
function createValidator(jdlObject, applicationSettings = {}, logger = console) {
    if (!jdlObject) {
        throw new Error('A JDL object must be passed to check for business errors.');
    }

    if (applicationSettings.blueprints && applicationSettings.blueprints.length !== 0) {
      return { checkForErrors: () => {} };
    } 

    return {
        checkForErrors: () => {
            checkForEntityErrors();
            checkForRelationshipErrors();
            checkForEnumErrors();
            checkDeploymentsErrors();
            checkForOptionErrors();
        },
    };

  // ...
}
```

That's the hardest part. The easiest part will be updating the `jdl-with-application-validation.js` file:
```javascript
function createValidator(jdlObject, logger = console) {
    if (!jdlObject) {
        throw new Error('A JDL object must be passed to check for business errors.');
    }

    return {
        checkForErrors: () => {
            checkForApplicationErrors();
            jdlObject.forEachApplication(jdlApplication => {
                const blueprints = jdlApplication.getConfigurationOptionValue('blueprints');
                if (!blueprints || blueprints.length > 0) {
                    return;
                }
                checkForEntityErrors(jdlApplication);
                checkForRelationshipErrors(jdlApplication);
                checkForEnumErrors();
                checkDeploymentsErrors();
                checkForOptionErrors(jdlApplication);
            });
            checkForRelationshipsBetweenApplications();
        },
    };

    // ...
}
```

There will be some testing to do: check no validation is performed when blueprints are passed in the two cases. This should be done at the highest level: for the `jdl-importer` file and for the two validators.

---

Hope it wasn't too long :) if you want to do this PR, it'd be awesome!

**Comment by nicolas63** [comment]
thank you @MathieuAA for your precious explanation, I will make the changes.

**Comment by mshima** [comment]
Print a warning:
```
      return { checkForErrors: () => {
          console.log('Generating application with blueprint, skipping jdl validation');
      } };
```

**Comment by MathieuAA** [comment]
Maybe not console.log, but console.warn :)

**Comment by nicolas63** [comment]
@MathieuAA 
`There will be some testing to do: check no validation is performed when blueprints are passed in the two cases. This should be done at the highest level: for the jdl-importer file and for the two validators.`

how can i do this ? i have try to spy logger but the method createValidator don't use
`const logger = require('../utils/objects/logger');`

should I use it rather than console in `function createValidator(jdlObject, logger = console) ` ?

**Comment by MathieuAA** [comment]
In the tests, pass whatever log object you want to test like it's done l.85 of the `test/jdl/validators/jdl-with-application-validator.spec.js` file. You can use spies too.

**Comment by nicolas63** [comment]
@MathieuAA i have added UT but some checks failed during CI, it's due to my modification ?

**Comment by MathieuAA** [comment]
I'll check right away

**Comment by MathieuAA** [comment]
It looks like an issue with Actions as a whole, I'm looking into it

File: jdl/jdl-importer.js [hunk 1]
```diff
@@ -102,9 +102,9 @@ function makeJDLImporter(content, configuration) {
          *          - exportedApplications: the exported applications, or an empty list
          *          - exportedEntities: the exported entities, or an empty list
          */
-        import: () => {
+        import: (logger = console) => {
```
Comment (MathieuAA) [inline jdl/jdl-importer.js:105]: you've forgotten to add the `logger` param

File: jdl/validators/jdl-without-application-validator.js [hunk 2]
```diff
@@ -46,13 +46,22 @@ module.exports = {
  * @param {String} applicationSettings.applicationType - the application type.
  * @param {String} applicationSettings.databaseType - the DB type.
  * @param {Boolean} applicationSettings.skippedUserManagement - whether user management is skipped.
+ * @param {Object} applicationSettings.blueprints - the blueprints used.
```
Comment (MathieuAA) [inline jdl/validators/jdl-without-application-validator.js:49]: Is it an array or an object? I'm seeing it responds to `length`

File: jdl/validators/jdl-without-application-validator.js [hunk 3]
```diff
@@ -46,13 +46,22 @@ module.exports = {
...
  */
 function createValidator(jdlObject, applicationSettings = {}, logger = console) {
     if (!jdlObject) {
         throw new Error('A JDL object must be passed to check for business errors.');
     }
 
+    if (applicationSettings.blueprints && applicationSettings.blueprints.length !== 0) {
+        return {
+            checkForErrors: () => {
+                logger.warn('Generating application with blueprint, skipping jdl validation');
```
Comment (MathieuAA) [inline jdl/validators/jdl-without-application-validator.js:60]: This message implies the validator knows about the context where it's being used, meaning the validator knows we're using it to generate applications. It shouldn't.
What about something like "Blueprints are being used, the JDL validation phase is skipped."?

File: jdl/validators/jdl-with-application-validator.js [hunk 4]
```diff
@@ -52,6 +52,11 @@ function createValidator(jdlObject, logger = console) {
         checkForErrors: () => {
             checkForApplicationErrors();
             jdlObject.forEachApplication(jdlApplication => {
+                const blueprints = jdlApplication.getConfigurationOptionValue('blueprints');
+                if (blueprints && blueprints.length > 0) {
+                    logger.warn('Generating application with blueprint, skipping jdl validation');
```
Comment (MathieuAA) [inline jdl/validators/jdl-with-application-validator.js:57]: This message implies the validator knows about the context where it's being used, meaning the validator knows we're using it to generate applications. It shouldn't.
What about something like "Blueprints are being used, the JDL validation phase is skipped."?

File: jdl/jdl-importer.js [hunk 5]
```diff
@@ -159,6 +159,7 @@ function checkForErrors(jdlObject, configuration) {
         let applicationType = configuration.applicationType;
         let databaseType = configuration.databaseType;
         let skippedUserManagement = configuration.skipUserManagement;
+        let blueprints = configuration.blueprints;
```
Comment (MathieuAA) [inline jdl/jdl-importer.js:162]: You've forgotten to add the "blueprints" param to the jsdoc starting from l.63.

File: test/jdl/jdl-importer.spec.js [hunk 6]
```diff
@@ -1863,6 +1869,10 @@ paginate * with infinite-scroll
                     },
                 ]);
             });
+
+            it('should not perform jdl validation', () => {
```
Comment (MathieuAA) [inline test/jdl/jdl-importer.spec.js:1873]:
```suggestion
            it('should warn about not performing jdl validation', () => {
```

File: test/jdl/validators/jdl-with-application-validator.spec.js [hunk 7]
```diff
@@ -675,5 +675,29 @@ describe('JDLWithApplicationValidator', () => {
...
+                const logger = {
+                    warn: callParameter => {
+                        parameter = callParameter;
+                    },
+                };
+                const validator = createValidator(jdlObject, logger);
+                validator.checkForErrors();
+            });
+
+            it('should not perform jdl validation', () => {
```
Comment (MathieuAA) [inline test/jdl/validators/jdl-with-application-validator.spec.js:698]:
```suggestion
            it('should warn about not performing jdl validation', () => {
```

File: test/jdl/validators/jdl-without-application-validator.spec.js [hunk 8]
```diff
@@ -521,5 +521,27 @@ describe('JDLWithoutApplicationValidator', () => {
...
+                };
+                const validator = createValidator(
+                    jdlObject,
+                    { blueprints: ['generator-jhipster-nodejs', 'generator-jhipster-dotnetcore'] },
+                    logger
+                );
+                validator.checkForErrors();
+            });
+
+            it('should not perform jdl validation', () => {
```
Comment (MathieuAA) [inline test/jdl/validators/jdl-without-application-validator.spec.js:542]:
```suggestion
            it('should warn about not performing jdl validation', () => {
```

**Review by MathieuAA — CHANGES_REQUESTED** [review]
Very nice :) only a few changes needed and it's good to go.

**Comment by MathieuAA** [comment]
> @MathieuAA i have added UT but some checks failed during CI, it's due to my modification ?

There was an issue with GH actions (I've noticed warning about 429 errors) and restarting the tests worked like a charm.

File: jdl/jdl-importer.js (hunk 1 above)
Comment (nicolas63) [reply jdl/jdl-importer.js:107]: done

File: jdl/validators/jdl-without-application-validator.js (hunk 2 above)
Comment (nicolas63) [reply jdl/validators/jdl-without-application-validator.js]: change to array

File: jdl/validators/jdl-without-application-validator.js (hunk 3 above)
Comment (nicolas63) [reply jdl/validators/jdl-without-application-validator.js]: ok with this

File: jdl/validators/jdl-with-application-validator.js (hunk 4 above)
Comment (nicolas63) [reply jdl/validators/jdl-with-application-validator.js]: ok with this

File: jdl/jdl-importer.js (hunk 5 above)
Comment (nicolas63) [reply jdl/jdl-importer.js:164]: added

File: test/jdl/jdl-importer.spec.js (hunk 6 above)
Comment (nicolas63) [reply test/jdl/jdl-importer.spec.js]: done

File: test/jdl/validators/jdl-without-application-validator.spec.js (hunk 8 above)
Comment (nicolas63) [reply test/jdl/validators/jdl-without-application-validator.spec.js]: done

File: test/jdl/validators/jdl-with-application-validator.spec.js (hunk 7 above)
Comment (nicolas63) [reply test/jdl/validators/jdl-with-application-validator.spec.js]: done

**Review by MathieuAA — APPROVED** [review]

**Comment by MathieuAA** [comment]
@nicolas63 thanks!! Feel free to merge it once the CI report goes green :)

**Comment by nicolas63** [comment]
thanks to you @MathieuAA for your advice ! Can you merge it ? I don't have the rights to do

**Pull request merged**